kernel/src/syscall.rs

```diff
@@ -1,104 +1,407 @@
 //! Tock syscall number definitions and arch-agnostic interface trait.
 
+use core::convert::TryFrom;
 use core::fmt::Write;
 
+use crate::driver::CommandReturn;
+use crate::errorcode::ErrorCode;
 use crate::process;
 
-/// The syscall number assignments.
+/// Helper function to split a u64 into a higher and lower u32.
+///
+/// Used in encoding 64-bit wide system call return values on 32-bit
+/// platforms.
+#[inline]
+fn u64_to_be_u32s(src: u64) -> (u32, u32) {
+    let src_bytes = src.to_be_bytes();
+    let src_msb = u32::from_be_bytes([src_bytes[0], src_bytes[1], src_bytes[2], src_bytes[3]]);
+    let src_lsb = u32::from_be_bytes([src_bytes[4], src_bytes[5], src_bytes[6], src_bytes[7]]);
+
+    (src_msb, src_lsb)
+}
+
+// ---------- SYSTEMCALL ARGUMENT DECODING ----------
+
+/// Enumeration of the system call classes based on the identifiers
+/// specified in the Tock ABI.
+#[repr(u8)]
+#[derive(Copy, Clone, Debug)]
+pub enum SyscallClass {
+    Yield = 0,
+    Subscribe = 1,
+    Command = 2,
+    ReadWriteAllow = 3,
+    ReadOnlyAllow = 4,
+    Memop = 5,
+    Exit = 6,
+}
+
+/// Enumeration of the yield system calls based on the Yield identifier
+/// values specified in the Tock ABI.
+#[repr(u8)]
+#[derive(Copy, Clone, Debug)]
+pub enum YieldCall {
+    NoWait = 0,
+    Wait = 1,
+}
+
+/// Enumeration of the exit system calls based on the Exit identifier values
+/// specified in the Tock ABI.
+#[repr(u8)]
```

Comment:
`u8`?

Reply:
The `ExitCall` type is not used anywhere in Tock. I propose we remove it.

Reply:
#2479